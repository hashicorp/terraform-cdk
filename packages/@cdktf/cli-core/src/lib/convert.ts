/**
 * Copyright (c) HashiCorp, Inc.
 * SPDX-License-Identifier: MPL-2.0
 */

import * as hcl2cdk from "@cdktf/hcl2cdk";
import { CdktfConfig } from "./cdktf-config";
import {
  ConstructsMakerProviderTarget,
  TerraformProviderConstraint,
  readSchema,
} from "@cdktf/provider-generator";
import { Errors, LANGUAGES } from "@cdktf/commons";

export async function convertConfigurationFile(configuration: string) {
  const cfg = CdktfConfig.read(process.cwd()); // TODO: make this the project directory instead of cwd
  const targets = cfg.terraformProviders.map((constraint) =>
    ConstructsMakerProviderTarget.from(
      new TerraformProviderConstraint(constraint),
      LANGUAGES[0]
    )
  );
  const { providerSchema } = await readSchema(targets);
  if (!providerSchema) {
    throw Errors.Internal("Could not fetch provider schema");
  }

  const { all } = await hcl2cdk.convert(configuration, {
    providerSchema,
    language: cfg.language,
    codeContainer: "constructs.Construct",
  });

  return all;
}
