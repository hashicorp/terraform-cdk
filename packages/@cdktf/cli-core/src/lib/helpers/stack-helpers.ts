// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0

import { ensureAllSettledBeforeThrowing, Errors, logger } from "@cdktf/commons";
import minimatch from "minimatch";
import { CdktfStack } from "../cdktf-stack";
import { SynthesizedStack } from "../synth-stack";

export function getSingleStack(
  stacks: SynthesizedStack[],
  stackName?: string,
  targetAction?: string,
) {
  if (!stacks) {
    throw Errors.Internal(
      "Trying to access a stack before it has been synthesized",
    );
  }

  if (stackName) {
    const stack = stacks.find((s) => s.name === stackName);
    if (!stack) {
      throw Errors.Usage("Could not find stack: " + stackName);
    }

    return stack;
  }

  if (stacks.length === 1) {
    return stacks[0];
  }

  throw Errors.Usage(
    `Found more than one stack, please specify a target stack. Run cdktf ${
      targetAction || "<verb>"
    } <stack> with one of these stacks: ${stacks
      .map((s) => s.name)
      .join(", ")} `,
  );
}

export function getMultipleStacks(
  stacks: SynthesizedStack[],
  patterns?: string[],
  targetAction?: string,
) {
  if (!patterns || !patterns.length) {
    if (stacks.length === 1) {
      return [stacks[0]];
    }
    throw Errors.Usage(
      `Found more than one stack, please specify a target stack. Run cdktf ${
        targetAction || "<verb>"
      } <stack> with one of these stacks: ${stacks
        .map((s) => s.name)
        .join(", ")} `,
    );
  }

  return patterns.flatMap((pattern) => {
    const matchingStacks = stacks.filter((stack) =>
      minimatch(stack.name, pattern),
    );

    if (matchingStacks.length === 0) {
      throw Errors.Usage(`Could not find stack for pattern '${pattern}'`);
    }

    return matchingStacks;
  });
}

// Returns the first stack that has no unmet dependencies
// An unmet dependency is a dependency that has not been deployed yet
// If there is no unfinished stack, returns undefined
// If there is no stack ready to be worked on, it returns a promise that will resolve as soon as there is a follow-up stack available
export async function getStackWithNoUnmetDependencies(
  stackExecutors: CdktfStack[],
): Promise<CdktfStack | undefined> {
  logger.debug("Checking for stacks with unmet dependencies");
  logger.debug("stack executors:", stackExecutors);
  const pendingStacks = stackExecutors.filter((executor) => executor.isPending);
  logger.debug("pending stacks:", stackExecutors);

  if (pendingStacks.length === 0) {
    return undefined;
  }

  const currentlyReadyStack = pendingStacks.find((executor) =>
    executor.stack.dependencies.every(
      (dependency) =>
        stackExecutors.find((executor) => executor.stack.name === dependency)
          ?.currentState === "done",
    ),
  );

  if (currentlyReadyStack) {
    logger.debug("Found a stack with no unmet dependencies");
    return currentlyReadyStack;
  }

  const stackExecutionPromises = stackExecutors
    .filter((ex) => ex.currentWorkPromise)
    .map((ex) => ex.currentWorkPromise);

  logger.debug(
    `${stackExecutionPromises.length} stacks are currently busy, waiting for one to finish`,
  );

  if (!stackExecutionPromises.length) {
    return undefined;
  }

  await ensureAllSettledBeforeThrowing(
    Promise.race(stackExecutionPromises),
    stackExecutionPromises,
  );

  return await getStackWithNoUnmetDependencies(stackExecutors);
}

function findAllDependantStacks(
  stackExecutors: CdktfStack[],
  stackName: string,
): CdktfStack[] {
  return stackExecutors.filter((innerExecutor) =>
    innerExecutor.stack.dependencies.includes(stackName),
  );
}

export function findAllNestedDependantStacks(
  stackExecutors: CdktfStack[],
  stackName: string,
  knownDependantStackNames: Set<string> = new Set(),
): CdktfStack[] {
  const dependantStacks = findAllDependantStacks(stackExecutors, stackName);
  dependantStacks.forEach((stack) => {
    if (knownDependantStackNames.has(stack.stack.name)) {
      return;
    }

    knownDependantStackNames.add(stack.stack.name);
    findAllNestedDependantStacks(
      stackExecutors,
      stack.stack.name,
      knownDependantStackNames,
    );
  });

  return stackExecutors.filter((executor) =>
    knownDependantStackNames.has(executor.stack.name),
  );
}

// Returns the first stack that has no dependents that need to be destroyed first
export async function getStackWithNoUnmetDependants(
  stackExecutors: CdktfStack[],
): Promise<CdktfStack | undefined> {
  logger.debug("Checking for stacks with unmet dependants");
  logger.debug("stack executors:", stackExecutors);
  const pendingStacks = stackExecutors.filter((executor) => executor.isPending);
  logger.debug("pending stacks:", stackExecutors);

  if (pendingStacks.length === 0) {
    return undefined;
  }

  const currentlyReadyStack = pendingStacks.find((executor) => {
    const dependantStacks = findAllDependantStacks(
      stackExecutors,
      executor.stack.name,
    );
    return dependantStacks.every((stack) => stack.currentState === "done");
  });

  if (currentlyReadyStack) {
    logger.debug("Found a stack with no unmet dependants");
    return currentlyReadyStack;
  }
  const stackExecutionPromises = stackExecutors
    .filter((ex) => ex.currentWorkPromise)
    .map((ex) => ex.currentWorkPromise);

  logger.debug(
    `${stackExecutionPromises.length} stacks are currently busy, waiting for one to finish`,
  );
  if (!stackExecutionPromises.length) {
    return undefined;
  }

  await Promise.race(stackExecutionPromises);
  return await getStackWithNoUnmetDependants(stackExecutors);
}

// Throws an error if there is a dependant stack that is not included
export function checkIfAllDependantsAreIncluded(
  stacksToRun: SynthesizedStack[],
  allStacks: SynthesizedStack[],
) {
  const allDependants = new Set<string>();
  stacksToRun
    .map((stack) =>
      allStacks.filter((s) => s.dependencies.includes(stack.name)),
    )
    .flat()
    .forEach((dependant) => allDependants.add(dependant.name));

  const stackNames = stacksToRun.map((stack) => stack.name);
  const missingDependants = [...allDependants].filter(
    (dependant) => !stackNames.includes(dependant),
  );

  if (missingDependants.length > 0) {
    throw Errors.Usage(
      `The following dependant stacks are not included in the stacks to run: ${missingDependants.join(
        ", ",
      )}. Either add them or add the --ignore-missing-stack-dependencies flag.`,
    );
  }
}

/** 
  Throws an error if there is a dependency that is not included
  Cycles are detected on dependency creation at synthesis time
  Running this prevents us from being in a situation where we 
  have to wait for a stack to be deployed that is not included to be run
*/
export function checkIfAllDependenciesAreIncluded(
  stacksToRun: SynthesizedStack[],
) {
  const allDependencies = new Set<string>();
  stacksToRun
    .map((stack) => stack.dependencies)
    .flat()
    .forEach((dependency) => allDependencies.add(dependency));

  const stackNames = stacksToRun.map((stack) => stack.name);
  const missingDependencies = [...allDependencies].filter(
    (dependency) => !stackNames.includes(dependency),
  );

  if (missingDependencies.length > 0) {
    throw Errors.Usage(
      `The following dependencies are not included in the stacks to run: ${missingDependencies.join(
        ", ",
      )}. Either add them or add the --ignore-missing-stack-dependencies flag.`,
    );
  }
}
