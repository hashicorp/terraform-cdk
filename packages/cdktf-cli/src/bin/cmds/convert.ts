// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import yargs from "yargs";
import { Errors } from "@cdktf/commons";
import { BaseCommand } from "./helper/base-command";
import { requireHandlers } from "./helper/utilities";
import * as path from "path";
import * as fs from "fs-extra";

function readCdktfJson(cwd = process.cwd()): { language: string } | undefined {
  try {
    const cdktfJsonPath = path.join(cwd, "cdktf.json");
    return fs.readJsonSync(cdktfJsonPath);
  } catch (e) {
    return undefined;
  }
}

class Command extends BaseCommand {
  public readonly command = "convert";
  public readonly describe =
    "Converts a single file of HCL configuration to CDK for Terraform. Takes the file to be converted on stdin.";

  public readonly builder = (args: yargs.Argv) =>
    args
      .example(
        "cat main.tf | cdktf convert  --provider integrations/github",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content and prints it"
      )
      .example(
        "cat main.tf | cdktf convert --provider hashicorp/aws > imported.ts",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content in imported.ts"
      )
      .example(
        "cat main.tf | cdktf convert --provider 'hashicorp/aws@ ~>3.62.0' 'integrations/github@ ~>4.16.0' --language python > imported.py",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content in imported.py"
      )
      .option("language", {
        choices: ["typescript", "python", "csharp", "java", "go"],
        default: readCdktfJson()?.language || "typescript",
      })
      .option("provider", {
        describe:
          "The conversion needs to know which providers are used in addition to the ones in your cdktf.json file. We search for a cdktf.json below your current working directory.",
        type: "array",
        default: [],
      })
      .option("stack", {
        describe: "Wrap the generated code within a stack class",
        type: "boolean",
        default: false,
      })
      .showHelpOnFail(true);

  public async handleCommand(argv: any) {
    Errors.setScope("convert");
    // deferred require to keep cdktf-cli main entrypoint small (e.g. for fast shell completions)
    const api = requireHandlers();
    if (argv.withProject !== undefined) {
      if (argv.language === "typescript") {
        argv.withProject = false;
      } else {
        argv.withProject = true;
      }
    }
    await api.convert(argv);
  }
}

module.exports = new Command();
