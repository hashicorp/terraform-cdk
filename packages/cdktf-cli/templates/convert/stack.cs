using System;
using Constructs;
using HashiCorp.Cdktf;
/* {{ imports }} */

namespace MyCompany.MyApp
{
    class MyStack : TerraformStack
    {
        public MyStack(Construct scope, string id) : base(scope, id)
        {
            /* {{ code }} */
        }
    }
}