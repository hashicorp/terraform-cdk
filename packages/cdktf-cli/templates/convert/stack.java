package com.mycompany.app;

import software.constructs.Construct;
import com.hashicorp.cdktf.TerraformStack;
/* {{ imports }} */

public class MyStack extends TerraformStack
{
    public MyStack(final Construct scope, final String id) {
        super(scope, id);

        /* {{ code }} */
    }
}