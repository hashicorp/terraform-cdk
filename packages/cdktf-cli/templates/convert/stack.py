#!/usr/bin/env python
# Copyright (c) HashiCorp, Inc.
# SPDX-License-Identifier: MPL-2.0

from constructs import Construct
from cdktf import TerraformStack
# {{ imports }}


class MyStack(TerraformStack):
    def __init__(self, scope: Construct, id: str):
        super().__init__(scope, id)

        # {{ code }}
