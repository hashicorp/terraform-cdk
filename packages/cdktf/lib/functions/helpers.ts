// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { Tokenization, Token } from "../tokens/token";
import { call } from "../tfExpression";
import { IResolvable } from "../tokens/resolvable";
import { TokenString, extractTokenDouble } from "../tokens/private/encoding";

type TFValue<T> = { variadic?: boolean; value: T };
type TFValueValidator<T> = (value: T) => TFValue<T>;

type ExecutableTfFunction = (...args: any[]) => IResolvable;

/**
 * Determines if given str has unescaped double quotes
 * @param str String to test
 * @returns Boolean
 */
function hasUnescapedDoubleQuotes(str: string) {
  return /(^|[^\\])([\\]{2})*"/.test(str);
}

// Validators
// eslint-disable-next-line jsdoc/require-jsdoc
export function anyValue<T>(value: T): TFValue<T> {
  return { value };
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function mapValue<T>(value: T): TFValue<T> {
  return { value };
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function stringValue<T extends string>(value: T): TFValue<T> {
  if (typeof value !== "string" && !Tokenization.isResolvable(value)) {
    throw new Error(`'${value}' is not a valid string nor a token`);
  }

  if (typeof value === "string" && hasUnescapedDoubleQuotes(value)) {
    throw new Error(
      `'${value}' can not be used as value directly since it has unescaped double quotes in it. To safely use the value please use Fn.rawString on your string.`
    );
  }

  return { value };
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function numericValue<T>(value: T): TFValue<T> {
  if (typeof value !== "number" && !Tokenization.isResolvable(value)) {
    throw new Error(`${value} is not a valid number nor a token`);
  }
  return { value };
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function listOf<T>(type: TFValueValidator<T>): TFValueValidator<T[]> {
  return (value: any) => {
    if (Tokenization.isResolvable(value)) {
      return { value };
    }

    if (!Array.isArray(value)) {
      //   throw new Error(`${value} is not a valid list`);
      return { value };
    }

    return {
      value: value
        .filter((item) => item !== undefined && item !== null)
        .map((item, i) => {
          if (Tokenization.isResolvable(item)) {
            return item;
          }

          if (TokenString.forListToken(item).test()) {
            return item;
          }

          if (extractTokenDouble(item, true) !== undefined) {
            return item;
          }

          if (TokenString.forMapToken(item).test()) {
            return item;
          }

          if (typeof item === "string") {
            const tokenList = Tokenization.reverseString(item);
            const numberOfTokens =
              tokenList.tokens.length + tokenList.intrinsic.length;
            if (numberOfTokens === 1 && tokenList.literals.length === 0) {
              return item;
            }
          }

          try {
            type(item);
            return typeof item === "string" ? `"${item}"` : item;
          } catch (error) {
            throw new Error(
              `Element in list ${value} at position ${i} is not of the right type: ${error}`
            );
          }
        }),
    };
  };
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function variadic<T>(type: TFValueValidator<T>): TFValueValidator<T[]> {
  return (value: any) => {
    // we use the list validator and set variadic to true in order to have it expanded in the args passed to the TF function
    return { value: listOf(type)(value).value, variadic: true };
  };
}

// Tokenization
// eslint-disable-next-line jsdoc/require-jsdoc
export function asString(value: IResolvable) {
  return Token.asString(value);
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function asNumber(value: IResolvable) {
  return Token.asNumber(value);
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function asList(value: IResolvable) {
  return Token.asList(value);
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function asStringMap(value: IResolvable) {
  return Token.asStringMap(value);
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function asBoolean(value: IResolvable) {
  return value; // Booleans can not be represented as a token
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function asAny(value: IResolvable) {
  // Ordinarily casting to any can cause issues, but
  // in this case it makes using functions a bit easier in TS
  // and doesn't really harm other languages.
  // Jsii has issues when returning the value directly,
  // so wrap as a string.
  return asString(value) as any;
}

// eslint-disable-next-line jsdoc/require-jsdoc
export function terraformFunction(
  name: string,
  argValidators: TFValueValidator<any>[]
): ExecutableTfFunction {
  return function (...args: any[]) {
    if (args.length !== argValidators.length) {
      throw new Error(
        `${name} takes ${argValidators.length} arguments, but ${args.length} were provided`
      );
    }
    return call(
      name,
      // We use flatMap now, since listOf() wraps everything in an extra array but variadic() does not
      // This enables us to have variadic pass multiple args into the call() from a single array
      args.reduce((carry, arg, i) => {
        try {
          const val = argValidators[i](arg);
          if (val.variadic) return [...carry, ...val.value];
          else return [...carry, val.value];
        } catch (error) {
          throw new Error(
            `Argument ${i} of ${name} failed the validation: ${error}`
          );
        }
      }, [])
    );
  };
}
