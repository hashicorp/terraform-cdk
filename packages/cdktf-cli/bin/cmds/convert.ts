import yargs from "yargs";
import { Errors } from "../../lib/errors";
import { requireHandlers } from "./helper/utilities";

class Command implements yargs.CommandModule {
  public readonly command = "convert [OPTIONS]";
  public readonly describe =
    "Converts a single file of HCL configuration to CDK for Terraform. Takes the file to be converted on stdin.";

  public readonly builder = (args: yargs.Argv) =>
    args
      .example(
        "cat main.tf | cdktf convert  --provider integrations/github",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content and prints it"
      )
      .example(
        "cat main.tf | cdktf convert --provider hashicorp/aws > imported.ts",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content in imported.ts"
      )
      .example(
        "cat main.tf | cdktf convert --provider 'hashicorp/aws@ ~>3.62.0' 'integrations/github@ ~>4.16.0' --language python > imported.py",
        "Takes the HCL content of main.tf and converts it to CDK for Terraform content in imported.ts"
      )
      .option("language", {
        choices: ["typescript", "python", "csharp", "java"],
        default: "typescript",
      })
      .option("provider", {
        describe:
          "The conversion needs to know which providers are used in addition to the ones in your cdktf.json file. We search for a cdktf.json below your current working directory.",
        type: "array",
        default: [],
      })
      .showHelpOnFail(true);

  public async handler(argv: any) {
    Errors.setScope("convert");
    // deferred require to keep cdktf-cli main entrypoint small (e.g. for fast shell completions)
    const api = requireHandlers();

    await api.convert(argv);
  }
}

module.exports = new Command();
