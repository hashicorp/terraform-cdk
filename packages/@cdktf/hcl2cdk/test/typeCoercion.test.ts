// Copyright (c) HashiCorp, Inc
// SPDX-License-Identifier: MPL-2.0
import { testCase, Synth, binding, Snapshot } from "./helpers/convert";

describe("type coercion", () => {
  testCase.test(
    "variables used in resources need to use stringValue",
    `
      provider "aws" {
        region                      = "us-east-1"
      }
      variable "zone_id" {
        type        = string
      }
      resource "aws_route53_record" "example_aws_route53_simple_record" {
        zone_id = var.zone_id
        name    = "example.com"
        type    = "TXT"
        ttl     = "300"
        records = ["example"]
      }
      `,
    [binding.aws],
    Snapshot.yes,
    Synth.yes,
    { resources: ["aws_route53_record"] }
  );

  testCase.test(
    "function arguments get coerced to the correct type",
    `
      provider "aws" {
        region     = "us-east-1"
      }
      variable "zone_id" {
        type       = string
      }
      resource "aws_route53_record" "example_aws_route53_simple_record" {
        zone_id = textencodebase64(var.zone_id, "UTF-16LE")
        name    = "example.com"
        type    = "TXT"
        ttl     = "300"
        records = ["example"]
      }
      `,
    [binding.aws],
    Snapshot.yes,
    Synth.yes,
    { resources: ["aws_route53_record"] }
  );

  testCase.test(
    "references used in resources need to use Token.asString",
    `
      provider "aws" {
        region                      = "us-east-1"
      }
      resource "aws_route53_zone" "example_aws_route53_simple_zone" {
        name = "example.com"
      }
      resource "aws_route53_record" "example_aws_route53_simple_record" {
        zone_id = aws_route53_zone.example_aws_route53_simple_zone.zone_id
        name    = "example.com"
        type    = "TXT"
        ttl     = "300"
        records = ["example"]
      }
      `,
    [binding.aws],
    Snapshot.yes,
    Synth.yes,
    { resources: ["aws_route53_zone", "aws_route53_record"] }
  );

  testCase.test(
    "references used in resource for_eachs need to understand how the value is used",
    `
      provider "aws" {
        region                      = "us-east-1"
      }
      variable "iam_groups_users" {
        type        = map(object({
          groups = list(string)
        }))
      }
      resource "aws_iam_user_group_membership" "iam_user_group_membership_user_groups" {
        for_each = var.iam_groups_users
        user     = each.key
        groups   = each.value["groups"]
      }
      `,
    [binding.aws],
    Snapshot.yes,
    Synth.yes,
    { resources: ["aws_iam_user_group_membership"] }
  );

  testCase.test(
    "expressions need to be coerced to required type",
    `
      provider "aws" {
        region                      = "us-east-1"
      }
      variable "example_spot_and_fargate_name" {
        type        = string
      }
      variable "role_arn" {
        type        = string
      }
      variable "private_subnets" {
        type        = list(object({ id = string }))
      }
      variable "public_subnets" {
        type        = list(object({ id = string }))
      }
      resource "aws_eks_cluster" "example_spot_and_fargate_eks_cluster" {
        name     = "$\{var.example_spot_and_fargate_name}"
        role_arn = var.role_arn

        vpc_config {
          subnet_ids = concat(var.private_subnets.*.id, var.public_subnets.*.id)
        }
      }
      `,
    [binding.aws],
    Snapshot.yes,
    Synth.yes,
    { resources: ["aws_eks_cluster"] }
  );

  testCase.test(
    "coerces string literal to number",
    `
      provider "aws" {
        region                      = "us-east-1"
      }

      resource "aws_security_group_rule" "client_alb_allow_outbound" {
        security_group_id = "sg-1234567890"
        type              = "egress"
        protocol          = -1
        from_port         = 0
        to_port           = 0
        cidr_blocks       = ["0.0.0.0/0"]
        ipv6_cidr_blocks  = ["::/0"]
        description       = "Allow any outbound traffic."
      }
      `,
    [binding.aws],
    Snapshot.yes,
    Synth.yes,
    { resources: ["aws_security_group_rule"] }
  );
});
