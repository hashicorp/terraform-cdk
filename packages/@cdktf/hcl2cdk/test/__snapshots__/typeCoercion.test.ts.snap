// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`type coercion coerces string literal to number snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { Op, Token, TerraformStack } from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { AwsProvider } from "./.gen/providers/aws/provider";
import { SecurityGroupRule } from "./.gen/providers/aws/security-group-rule";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });
    new SecurityGroupRule(this, "client_alb_allow_outbound", {
      cidrBlocks: ["0.0.0.0/0"],
      description: "Allow any outbound traffic.",
      fromPort: 0,
      ipv6CidrBlocks: ["::/0"],
      protocol: Token.asString(Op.negate(1)),
      securityGroupId: "sg-1234567890",
      toPort: 0,
      type: "egress",
    });
  }
}
"
`;

exports[`type coercion expressions need to be coerced to required type snapshot typescript 1`] = `
"import { Construct } from "constructs";
import {
  VariableType,
  TerraformVariable,
  Fn,
  Token,
  TerraformStack,
} from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { EksCluster } from "./.gen/providers/aws/eks-cluster";
import { AwsProvider } from "./.gen/providers/aws/provider";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });
    const exampleSpotAndFargateName = new TerraformVariable(
      this,
      "example_spot_and_fargate_name",
      {
        type: VariableType.STRING,
      }
    );
    const privateSubnets = new TerraformVariable(this, "private_subnets", {
      type: VariableType.list(
        VariableType.object({
          id: VariableType.STRING,
        })
      ),
    });
    const publicSubnets = new TerraformVariable(this, "public_subnets", {
      type: VariableType.list(
        VariableType.object({
          id: VariableType.STRING,
        })
      ),
    });
    const roleArn = new TerraformVariable(this, "role_arn", {
      type: VariableType.STRING,
    });
    new EksCluster(this, "example_spot_and_fargate_eks_cluster", {
      name: exampleSpotAndFargateName.stringValue,
      roleArn: roleArn.stringValue,
      vpcConfig: {
        subnetIds: Token.asList(
          Fn.concat([
            Fn.lookupNested(privateSubnets.value, ["*", "id"]),
            Fn.lookupNested(publicSubnets.value, ["*", "id"]),
          ])
        ),
      },
    });
  }
}
"
`;

exports[`type coercion function arguments get coerced to the correct type snapshot typescript 1`] = `
"import { Construct } from "constructs";
import {
  VariableType,
  TerraformVariable,
  Token,
  Fn,
  TerraformStack,
} from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { AwsProvider } from "./.gen/providers/aws/provider";
import { Route53Record } from "./.gen/providers/aws/route53-record";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });
    const zoneId = new TerraformVariable(this, "zone_id", {
      type: VariableType.STRING,
    });
    new Route53Record(this, "example_aws_route53_simple_record", {
      name: "example.com",
      records: ["example"],
      ttl: Token.asNumber("300"),
      type: "TXT",
      zoneId: Token.asString(
        Fn.textencodebase64(zoneId.stringValue, "UTF-16LE")
      ),
    });
  }
}
"
`;

exports[`type coercion references used in resource for_eachs need to understand how the value is used snapshot typescript 1`] = `
"import { Construct } from "constructs";
import {
  VariableType,
  TerraformVariable,
  Token,
  TerraformIterator,
  Fn,
  TerraformStack,
} from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { IamUserGroupMembership } from "./.gen/providers/aws/iam-user-group-membership";
import { AwsProvider } from "./.gen/providers/aws/provider";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });
    const iamGroupsUsers = new TerraformVariable(this, "iam_groups_users", {
      type: VariableType.map(
        VariableType.object({
          groups: VariableType.ANY,
        })
      ),
    });
    /*In most cases loops should be handled in the programming language context and 
    not inside of the Terraform context. If you are looping over something external, e.g. a variable or a file input
    you should consider using a for loop. If you are looping over something only known to Terraform, e.g. a result of a data source
    you need to keep this like it is.*/
    const iamUserGroupMembershipUserGroupsForEachIterator =
      TerraformIterator.fromList(Token.asAny(iamGroupsUsers.value));
    new IamUserGroupMembership(this, "iam_user_group_membership_user_groups", {
      groups: Token.asList(
        Fn.lookupNested(iamUserGroupMembershipUserGroupsForEachIterator.value, [
          '["groups"]',
        ])
      ),
      user: Token.asString(iamUserGroupMembershipUserGroupsForEachIterator.key),
      forEach: iamUserGroupMembershipUserGroupsForEachIterator,
    });
  }
}
"
`;

exports[`type coercion references used in resources need to use Token.asString snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { Token, TerraformStack } from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { AwsProvider } from "./.gen/providers/aws/provider";
import { Route53Record } from "./.gen/providers/aws/route53-record";
import { Route53Zone } from "./.gen/providers/aws/route53-zone";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });
    const exampleAwsRoute53SimpleZone = new Route53Zone(
      this,
      "example_aws_route53_simple_zone",
      {
        name: "example.com",
      }
    );
    new Route53Record(this, "example_aws_route53_simple_record", {
      name: "example.com",
      records: ["example"],
      ttl: Token.asNumber("300"),
      type: "TXT",
      zoneId: exampleAwsRoute53SimpleZone.zoneId,
    });
  }
}
"
`;

exports[`type coercion variables used in resources need to use stringValue snapshot typescript 1`] = `
"import { Construct } from "constructs";
import { VariableType, TerraformVariable, Token, TerraformStack } from "cdktf";
/*
 * Provider bindings are generated by running \`cdktf get\`.
 * See https://cdk.tf/provider-generation for more details.
 */
import { AwsProvider } from "./.gen/providers/aws/provider";
import { Route53Record } from "./.gen/providers/aws/route53-record";
class MyConvertedCode extends TerraformStack {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    /*Terraform Variables are not always the best fit for getting inputs in the context of Terraform CDK.
    You can read more about this at https://cdk.tf/variables*/
    new AwsProvider(this, "aws", {
      region: "us-east-1",
    });
    const zoneId = new TerraformVariable(this, "zone_id", {
      type: VariableType.STRING,
    });
    new Route53Record(this, "example_aws_route53_simple_record", {
      name: "example.com",
      records: ["example"],
      ttl: Token.asNumber("300"),
      type: "TXT",
      zoneId: zoneId.stringValue,
    });
  }
}
"
`;
